### PR TITLE
Fix a deadlock in syscallbuf unmapping after vfork

### PR DESCRIPTION
**Pull request by KJTsanaktsidis**
This should be :crossed_fingers: a fix for https://github.com/rr-debugger/rr/issues/3807

It does two things to address that issue:

Firstly, moves the code for doing syscallbuf unmap after `vfork`/`exec` out of `Task::post_exec`, and instead defers doing that until we next run a task in that address space. This means we should never leak address space just because there were no available threads to perform the unmapping. This doesn't actually do anything to technically fix the issue, but it means we can write some reliable tests around the syscallbuf unmapping stuff, so it seemed worthwhile.

Secondly, we now look at the desched signal state in AutoRemoteSyscalls. If the desched signal is armed, we temporarily disarm it (and mask it too, in case the signal was pending but unprocessed). This is what should fix our issue, I think

It's hard to prove that this really has fixed the issue, but I left my reproduction script running for 12 hours last night and it didn't hang, and that had always been long enough to trigger the hang before. So I guess either this fixes the issue, or I got (un)lucky.

I'm sure my C++ isn't going to win any beauty contests here but hopefully this demonstrates the outlines of a fix.

### CODE REVIEW

File: src/Task.cc [hunk 1]
```diff
@@ -1130,6 +1118,21 @@ void Task::post_exec_syscall(const std::string& original_exe_file) {
 
 bool Task::execed() const { return tg->execed; }
 
+void Task::unmap_dead_syscallbufs_if_required() {
+  if (!as->regions_pending_unmap.empty()) {
+    LOG(warn) << "Using " << tid << " to unmap syscallbuf regions for "
+              << "previously exec'd processes";
+    AutoRemoteSyscalls remote(this);
+    std::vector<MemoryRange> regions_pending_unmap;
+    std::swap(regions_pending_unmap, as->regions_pending_unmap);
```
Comment (rocallahan) [inline src/Task.cc:1127]: This could be a move, but it's fine.

**Review by rocallahan — APPROVED** [review]
This is really excellent work in a very difficult area. Bravo!

**Comment by rocallahan** [comment]
Honestly, really extraordinary work.

**Comment by KJTsanaktsidis** [comment]
Thank you! I really appreciated your pointing in the right direction. This was definitely a fun exercise and I learned quite a bit about how rr works. And brushed off a bit of my C++ too.

**Pull request merged**